 0016-capi-upgrades/0016-capi-upgrades.md

```diff
@@ -0,0 +1,49 @@
+# RFC 016 - CLUSTER API UPGRADES
+
+## Reasoning
+
+As part of the Cluster API (CAPI) Hive Sprint, the `upgrades hive` decided to create a document to gather all actions need to be done in order to upgrade a workload cluster successfully as well as bring the functionality we have already built to trigger the upgrades.
+
+## Scope
+
+This document covers only the action to upgrade the workload cluster without solving problems related to the versioning of the different CAPI controllers and/or deciding which controller reconciles a cluster.
+
+At the same time all the automation on top of the upgrades, like freeze periods or maintainance windows, are out of the scope here ([include in RFC 0015](/0015-automatic-cluster--upgrades)).
+
+Furthermore, a workload cluster can be upgraded in different ways:
+
+- Apps installed in the cluster are upgraded (out of scope here). We have other non-CAPI components that already take care of that, like app-operator
+
+- Cluster is reconciled by a newer CAPI controller. We may want to upgrade the Cluster to a newer CAPI controller release, which may change the infrastructure created to support the cluster (network, load balancers, etc). We do this update just changing the CAPI label on the Cluster CR, and we don't need a new machine image to be created/referenced by the CR's.
+
+- Machine image used by control plane or workers is updated with new k8s version or new OS version. We may want to update the k8s version or the OS version running in the machines. In both cases, we need a new machine image using the updated components, and reference that image from the CR's.
+
+## Context
+
+The current approach to upgrade a workload cluster in CAPI world involves a series of steps described below:
```

Comment:
The approach below applies only to actual image changes. However, other changes like e.g. a Kubernetes version change are not as complex. In the mentioned K8s upgrade case it is just changing a spec field in the KubeadmControlPlane and then in each of the MachinePools/MachineDeployments.

Reply:
but kubectl and other components baked in the image needs to be updated too, right?

Reply:
the thing is that with k8s version changes you do not need to touch anything but the MachineTemplate, at least when I tried it I only needed to change the version field in CP and MachineDeployments, no changing of machine templates, so I'm not sure if that step is needed